### PR TITLE
feat: restrict discrepancies list to admin's DBCs matching in TCS or GMC

### PR DESCRIPTION
**Pull request by CaiWillis**
TIS21-4633: Reval Connections MVP: Restrict Discrepancies list by Designated Body

### CODE REVIEW

**Comment by CaiWillis** [comment]
Some notes:

- We already pass the Admin's DBCs from the front end for this endpoint
- The es query can roughly be read as:
`NOT Military
AND
   (
      param 1 matches designatedBody 
      OR 
      param 1 matches tcsDesignatedBody
   )
AND
   (
      param 0 matches doctorFirstName
      OR
      param 0 matches doctorLastName
      OR
      param 0 matches gmcNumber
   )`

File: src/main/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelper.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  /**
+   * Format designated body codes for Elasticsearch by stripping hyphens and setting lower case.
+   *
+   * @param designatedBodyCodes list of designated body codes to format
+   *
+   */
+  public static String formatDesignatedBodyCodesForElasticsearchQuery(
+      List<String> designatedBodyCodes) {
+    List<String> escapedCodes = new ArrayList<>();
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelper.java:41]: If you have some spare time, you might get a kick out of using a java stream from `designatedBodyCodes`, with map and "joining" collector, to make this redundant.

**Comment by hihilary** [comment]
Had a Postman test on my local, the dbc search query works.

However, I've just noticed `DiscrepanciesElasticSearchService` and `ConnectedElasticSearchService` look very similar, but one throws a RuntimeException, the other throws a self-defined exception. And I'm also wondering if it's worth refactoring these 2 services to reuse the same `searchForPage` method.

File: src/main/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelper.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.tis.revalidation.connection.service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ElasticsearchQueryHelper {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelper.java:27]: As a Util class, we should mark it as `final`

**Review by hihilary — APPROVED** [review]
I have questions in the previous comment. The PR itself looks good to me.

**Comment by CaiWillis** [comment]
> Had a Postman test on my local, the dbc search query works.
> 
> However, I've just noticed `DiscrepanciesElasticSearchService` and `ConnectedElasticSearchService` look very similar, but one throws a RuntimeException, the other throws a self-defined exception. And I'm also wondering if it's worth refactoring these 2 services to reuse the same `searchForPage` method.

Yep I'll sort the exceptions out to make everything tidy, and will have a think about them using the same method. I think to begin with they looked very different but have become similar, so it's whether they're likely to be different again

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelperTest.java
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelperTest.java:1]: Is there value in having test of empty & null values and values without `1-`?

**Review by JosephKelly — COMMENTED** [review]
I think the practice of making Util classes `final` isn't as essential but still has value in declaring intent

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelperTest.java
Comment (CaiWillis) [reply src/test/java/uk/nhs/hee/tis/revalidation/connection/service/ElasticsearchQueryHelperTest.java:1]: I don't think so(?) The method will be passed a list of strings from the front end, if there was something weird in that list like List.of("validString", "", null "") then something else has gone wrong

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/service/DiscrepanciesElasticSearchServiceTest.java [hunk 3]
```diff
@@ -117,21 +122,47 @@ void shouldSearchForPage() throws ConnectionQueryException {
...
+    assertThat(result, is(discrepanciesSummary));
+  }
+
   @Test
   void shouldThrowRuntimeExceptionWhenSearchForPage() {
     String searchQuery = gmcRef1;
     final var pageableAndSortable = PageRequest.of(Integer.parseInt(PAGE_NUMBER_VALUE), 20,
         by(ASC, "gmcReferenceNumber"));
+    final List<String> dbcs = List.of(designatedBody1, designatedBody2);
+    final String formattedDbcs = "1rsspz7 1rssq1b";
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/connection/service/DiscrepanciesElasticSearchServiceTest.java:159]: You should define this once in the class.

File: src/test/java/uk/nhs/hee/tis/revalidation/connection/service/DiscrepanciesElasticSearchServiceTest.java [hunk 4]
```diff
@@ -105,8 +107,11 @@ public void setup() {
   void shouldSearchForPage() throws ConnectionQueryException {
     final var pageableAndSortable = PageRequest.of(Integer.parseInt(PAGE_NUMBER_VALUE), 20,
         by(ASC, "gmcReferenceNumber"));
+    final List<String> dbcs = List.of(designatedBody1, designatedBody2);
+    final String formattedDbcs = "1rsspz7 1rssq1b";
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/connection/service/DiscrepanciesElasticSearchServiceTest.java:111]: You should define these once in the class.

**Review by JosephKelly — CHANGES_REQUESTED** [review]
Minor tweaks that shouldn't affect it passing.

**Review by JosephKelly — APPROVED** [review]
Minor tweaks that shouldn't affect it passing.

**Pull request merged**